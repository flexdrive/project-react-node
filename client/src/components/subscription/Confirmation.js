import React from 'react';
import formatters from '../../utilities/fomatters';
import vehicleEvergreen from '../../assets/images/vehicle_evergreen.png';
import constants from "../../constants";
import navigation from "../../utilities/navigation";


const Confirmation = ({subscription}) => (
    <div id="subscription-container">
        <h1>Subscription Summary</h1>
        <div id>
            Thank you for choosing FlexDrive. You're now in the drivers seat.
        </div>
        <div id="subscription-summary">
            <div id="customer-info">
                <h2>Customer Information</h2>
                <ul className="summary">
                    <li>
                        <div className="key">Name</div>
                        <div className="key-value">{subscription.customer.name}</div>
                    </li>
                    <li>
                        <div className="key">Email Address</div>
                        <div className="key-value">{subscription.customer.emailAddress}</div>
                    </li>
                    <li>
                        <div className="key">Birthday</div>
                        <div className="key-value">{subscription.customer.dateOfBirth}</div>
                    </li>
                </ul>
            </div>
            <div id="subscription-length-info">
                <h2>Subscription Information</h2>
                <ul className="summary">
                    <li>
                        <div className="key">Subscription Length</div>
                        <div className="key-value">{formatters.daysToWeeks(subscription.subscriptionLength.length)}</div>
                    </li>
                    <li>
                        <div className="key">Price</div>
                        <div className="key-value price">{formatters.numberToCurrency(subscription.subscriptionLength.price)}</div>
                    </li>
                </ul>
            </div>
            <div id="vehicle-info">
                <h2>Vehicle Information</h2>
                <ul className="summary">
                    <li>
                        <div className="key">Make</div>
                        <div className="key-value">{subscription.vehicle.make}</div>
                    </li>
                    <li>
                        <div className="key">Model</div>
                        <div className="key-value">{subscription.vehicle.model}</div>
                    </li>
                    <li>
                        <div className="key">Model Year</div>
                        <div className="key-value">{subscription.vehicle.year}</div>
                    </li>
                    <li>
                        <div className="key">Milage</div>
                        <div className="key-value">{subscription.vehicle.odometer}</div>
                    </li>
                    <li>
                        <div className="key">VIN ID</div>
                        <div className="key-value">{subscription.vehicle.vin}</div>
                    </li>
                </ul>
            </div>
            <img className="vehicle-image" src={vehicleEvergreen} alt={`${subscription.vehicle.make} ${subscription.vehicle.model}`} />
            <button className="button-success" onClick={() => navigation.go(constants.ROUTES.HOME)}>Home</button>
        </div>
    </div>
);

export default Confirmation;