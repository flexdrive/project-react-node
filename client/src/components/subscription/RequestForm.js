import React, {Component} from 'react';
import SubscriptionLengthModel from 'project-react-node-resources/lib/models/subscriptionLength';
import formatters from '../../utilities/fomatters';
import navigation from '../../utilities/navigation';
import formDataToJson from '../../utilities/formDataToJson';
import constants from '../../constants';
import vehicleEvergreen from '../../assets/images/vehicle_evergreen.png';
import subscriptionActions from '../../store/subscriptions/actions';
import subscriptionPropTypes from "../../store/subscriptionPropTypes";


class RequestForm extends Component{

    static propTypes = subscriptionPropTypes;

    constructor(props){
        super(props);

        this.state = {
            name: '',
            emailAddress: '',
            dateOfBirth: '',
            vehicle: this.props.subscriptionStore.vehicleList[0],
            subscriptionLength: new SubscriptionLengthModel(this.props.subscriptionStore.subscriptionLengthList[0].length)
        }
    }

    componentWillReceiveProps(nextProps){
        if(this.props.subscriptionStore.subscriptionLength.price !== nextProps.subscriptionStore.subscriptionLength.price)
            this.setState({subscriptionLength: nextProps.subscriptionStore.subscriptionLength});

        if(nextProps.subscriptionStore.subscription.customer !== null)
            navigation.go(constants.ROUTES.SUBSCRIPTION_CONFIRMATION);
    }

    render(){
        return (
            <React.Fragment>
                <form id="request-form" onSubmit={this.formSubmissionHandler}>
                    <div className="entry-field">
                        <input id="name" name="name"
                               className="effect-16"
                               type="text"
                               value={this.state.value}
                               onChange={(e) => this.changeHandlerFields('name', e)}
                               placeholder=""
                               required />
                        <label>Name</label>
                        <span className="focus-border"/>
                    </div>
                    <div className="entry-field">
                        <input id="emailAddress" name="emailAddress"
                               className="effect-16"
                               type="email"
                               value={this.state.value}
                               onChange={(e) => this.changeHandlerFields('emailAddress', e)}
                               placeholder=""
                               required />
                        <label>Email</label>
                        <span className="focus-border"/>
                    </div>
                    <div className="entry-field">
                        <input id="dateOfBirth" name="dateOfBirth"
                               className="effect-16"
                               type="date"
                               value={this.state.dateOfBirth}
                               onChange={(e) => this.changeHandlerFields('dateOfBirth', e)}
                               placeholder="Date"
                               required />
                        <label>Birthday</label>
                        <span className="focus-border"/>
                    </div>
                    {this.subscriptionLengthControlMarkup()}
                    {this.vehicleControlMarkup()}
                    {this.vehicleDetailMarkup()}
                    <div className="button-wrapper">
                        <input className="button-success" type="submit" value="Submit" />
                        <button className="button-cancel" onClick={() => navigation.go(constants.ROUTES.HOME)}>Cancel</button>
                    </div>
                </form>
            </React.Fragment>
        );
    }

    vehicleControlMarkup = () => {
        return (
            <React.Fragment>
                <div className="entry-field">
                    <label>Vehicle</label>
                    <select id="vehicle" name="vehicle" value={this.state.vehicle && this.state.vehicle.vin} onChange={this.changeHandlerVehicle}>
                        {this.props.subscriptionStore.vehicleList.map((vehicle, index) => {
                            return <option key={`vl_${index}`} value={vehicle.vin}>{vehicle.make} {vehicle.model}</option>;
                        })}
                    </select>
                </div>
            </React.Fragment>
        );
    }

    subscriptionLengthControlMarkup = () => {
        return (
            <React.Fragment>
                <div className="entry-field">
                    <label>Subscription Length</label>
                    <select id="subscriptionLength" name="subscriptionLength" value={this.state.subscriptionLength && this.state.subscriptionLength.length} onChange={this.changeHandlerSubscriptionLength}>
                        {this.props.subscriptionStore.subscriptionLengthList.map((listItem, index) => {
                            return <option key={`sl_${index}`} value={listItem.length}>{formatters.daysToWeeks(listItem.length)}</option>;
                        })}
                    </select>
                </div>
            </React.Fragment>
        );
    }

    vehicleDetailMarkup(){
        let vehicle = this.state.vehicle;
        if(vehicle === '')
            vehicle = this.props.subscriptionStore.vehicleList[0];

        return (
            <div id="vehicle-info">
                <h2>Vehicle Information</h2>
                <ul className="summary">
                    <li>
                        <div className="key">Make</div>
                        <div className="key-value">{vehicle.make}</div>
                    </li>
                    <li>
                        <div className="key">Model</div>
                        <div className="key-value">{vehicle.model}</div>
                    </li>
                    <li>
                        <div className="key">Model Year</div>
                        <div className="key-value">{vehicle.year}</div>
                    </li>
                    <li>
                        <div className="key">Milage</div>
                        <div className="key-value">{vehicle.odometer}</div>
                    </li>
                    <li>
                        <div className="key">VIN ID</div>
                        <div className="key-value">{vehicle.vin}</div>
                    </li>
                    <li>
                        <div className="key">PRICE</div>
                        <div className="key-value price">{formatters.numberToCurrency(this.state.subscriptionLength.price)}</div>
                    </li>
                </ul>
                <img className="vehicle-image" src={vehicleEvergreen} alt={`${vehicle.make} ${vehicle.model}`} />
            </div>
        )
    }

    changeHandlerFields = (field, event) => {
        const selectedValue = event.target.value;

        this.setState({[field]: selectedValue});
    }

    changeHandlerSubscriptionLength = (event) => {
        const selectedValue = event.target.value;

        /*
        * Get Price info from micro service
        * */
        this.props.actions.dispatch(()=> subscriptionActions.getSubscriptionPrice(selectedValue, this.state.vehicle.vin));
    }

    changeHandlerVehicle = (event) => {
        const selectedValue = event.target.value;
        const selectedVehicle = this.props.subscriptionStore.vehicleList.filter(vehicle => vehicle.vin === selectedValue);

        if(selectedVehicle.length)
            this.setState({vehicle: selectedVehicle[0]});
    }

    formSubmissionHandler = (event) => {
        event.preventDefault();
        const formData = new FormData(event.target);

        /*
        * convert the form data object to a usable format and
        * update the form values with the appropriate model for persistence
        * */
       const jsonForm = formDataToJson(formData);
       const completedFormModel = Object.assign({}, JSON.parse(jsonForm), {
           subscriptionLength: this.state.subscriptionLength,
           vehicle: this.state.vehicle
       });

        /*
        * persist subscription - could use a then-able and ASSUME persistence, but will use props eval to validate
        * */
        this.props.actions.dispatch(()=> subscriptionActions.setSubscription(JSON.stringify(completedFormModel)))
    }
}

export default RequestForm;