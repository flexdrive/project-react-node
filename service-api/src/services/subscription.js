import { send, json, text } from 'micro';
import microCors from 'micro-cors';
import { get, post, withNamespace } from 'microrouter';
import CustomerModel from 'project-react-node-resources/lib/models/customer';


const cors = microCors({ allowMethods: ['GET', 'POST'], allowHeaders: ['Access-Control-Allow-Origin','Content-Type','Accept'] });

const subscription = withNamespace('/api/subscription')(
    get('/', () => 'My subscription route'),
    post('/', async (req, res) => {
        const subscription = await json(req);
        const customerModel = new CustomerModel();
        customerModel.mapFromService(subscription);

        /*
        * TODO: implement a storage mechanism
        * Right now just updating the subscription model with a customer
        * and passing back through for reference in local state store for confirmation
        * The intent is demonstrated
        * */
        const response = Object.assign({}, subscription, {customer: customerModel});
        console.log('Updated_Subscription', response);

        return response;
    })
);

export default cors(subscription);