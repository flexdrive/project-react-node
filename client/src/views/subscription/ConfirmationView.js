import React, {Component} from 'react';
import Confirmation from '../../components/subscription/Confirmation';
import navigation from "../../utilities/navigation";
import constants from '../../constants';
import connectedStoreContainer from '../../store/connectedStoreContainer';
import subscriptionPropTypes from '../../store/subscriptionPropTypes';


class ConfirmationView extends Component{

    static propTypes = subscriptionPropTypes;

    constructor(props){
        super(props);
    }

    componentDidMount() {
        if(!this.props.subscriptionStore.subscriptionLengthList.length)
            navigation.go(constants.ROUTES.HOME);
    }

    render(){
        if(!this.props.subscriptionStore.subscriptionLengthList.length)
            return null;

        return (
            <Confirmation subscription={this.props.subscriptionStore.subscription} />
        );
    }
}

export default connectedStoreContainer(ConfirmationView);