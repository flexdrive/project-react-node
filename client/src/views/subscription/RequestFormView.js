import React, {Component} from 'react';
import RequestForm from '../../components/subscription/RequestForm';
import connectedStoreContainer from '../../store/connectedStoreContainer';
import subscriptionPropTypes from '../../store/subscriptionPropTypes';
import navigation from '../../utilities/navigation';
import constants from '../../constants';


class RequestFormView extends Component{

    static propTypes = subscriptionPropTypes;

    constructor(props){
        super(props);
    }

    componentDidMount(){
        if(!this.props.subscriptionStore.subscriptionLengthList.length)
            navigation.go(constants.ROUTES.HOME);
    }

    render(){
        if(!this.props.subscriptionStore.subscriptionLengthList.length)
            return null;

        return (
            <RequestForm {...this.props}/>
        );
    }
}


export default connectedStoreContainer(RequestFormView);