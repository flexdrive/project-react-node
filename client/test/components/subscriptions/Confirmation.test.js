import  React from 'react';
import {subscriptionMock} from "../../mocks/subscription.mock";
import Confirmation from '../../../src/components/subscription/Confirmation';


describe('Subscription Confirmation Component test suite', () => {
    let props;

    beforeEach(() =>{
        props = {subscription: subscriptionMock};
    });

    afterEach(() => {
        props = null;
    });

    describe('Unit Testing: Render', () => {
        test('evalutate render and props', () => {
            const component = new Confirmation(props);

            React.isValidElement(component);
            expect(component.props.children[2].props.id).toEqual('subscription-summary');
        });
    });

    describe('Component Testing: Render', () => {
        test('should render component', () => {
            const element = shallow(<Confirmation {...props} />);

            expect(element.find('#subscription-container').length).toBe(1);  //parent div element
            expect(element.find('.vehicle-image').length).toBe(1);           //image element
            expect(element).toMatchSnapshot();
        });
    });
});
