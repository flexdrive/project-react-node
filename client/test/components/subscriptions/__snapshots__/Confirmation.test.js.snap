// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscription Confirmation Component test suite Component Testing: Render should render component 1`] = `
<div
  id="subscription-container"
>
  <h1>
    Subscription Summary
  </h1>
  <div
    id={true}
  >
    Thank you for choosing FlexDrive. You're now in the drivers seat.
  </div>
  <div
    id="subscription-summary"
  >
    <div
      id="customer-info"
    >
      <h2>
        Customer Information
      </h2>
      <ul
        className="summary"
      >
        <li>
          <div
            className="key"
          >
            Name
          </div>
          <div
            className="key-value"
          >
            John Smith
          </div>
        </li>
        <li>
          <div
            className="key"
          >
            Email Address
          </div>
          <div
            className="key-value"
          >
            jsmith@gmail.com
          </div>
        </li>
        <li>
          <div
            className="key"
          >
            Birthday
          </div>
          <div
            className="key-value"
          >
            10/01/1990
          </div>
        </li>
      </ul>
    </div>
    <div
      id="subscription-length-info"
    >
      <h2>
        Subscription Information
      </h2>
      <ul
        className="summary"
      >
        <li>
          <div
            className="key"
          >
            Subscription Length
          </div>
          <div
            className="key-value"
          >
            4 weeks
          </div>
        </li>
        <li>
          <div
            className="key"
          >
            Price
          </div>
          <div
            className="key-value price"
          >
            $750
          </div>
        </li>
      </ul>
    </div>
    <div
      id="vehicle-info"
    >
      <h2>
        Vehicle Information
      </h2>
      <ul
        className="summary"
      >
        <li>
          <div
            className="key"
          >
            Make
          </div>
          <div
            className="key-value"
          >
            Audi
          </div>
        </li>
        <li>
          <div
            className="key"
          >
            Model
          </div>
          <div
            className="key-value"
          >
            A4
          </div>
        </li>
        <li>
          <div
            className="key"
          >
            Model Year
          </div>
          <div
            className="key-value"
          >
            2017
          </div>
        </li>
        <li>
          <div
            className="key"
          >
            Milage
          </div>
          <div
            className="key-value"
          >
            18200
          </div>
        </li>
        <li>
          <div
            className="key"
          >
            VIN ID
          </div>
          <div
            className="key-value"
          >
            1ABCD23HIJK112233
          </div>
        </li>
      </ul>
    </div>
    <img
      alt="Audi A4"
      className="vehicle-image"
      src={
        Object {
          "process": [Function],
        }
      }
    />
  </div>
</div>
`;
